lib/api/multiObjectDelete.js

```diff
@@ -0,0 +1,346 @@
+*/
+function _formatXML(quietSetting, errorResults, deleted) {
+    let errorXML = [];
+    errorResults.forEach(errorObj => {
+        errorXML.push(
+        '<Error>',
+        '<Key>', escapeForXML(errorObj.key), '</Key>',
+        '<Code>', errorObj.error.message, '</Code>',
+        '<Message>', errorObj.error.description, '</Message>',
+        '</Error>'
+        );
+    });
+    errorXML = errorXML.join('');
+    const xml = [
+        '<DeleteResult ',
+        'xmlns="http://s3.amazonaws.com/doc/2006-03-01/">',
+        // placeholder in array for deleted list if verbose response
+        '',
+        errorXML,
+        '</DeleteResult>',
+    ];
+    if (quietSetting) {
+        // just return errors
+        return xml.join('');
+    }
+    const deletedXML = [];
+    deleted.forEach(objKey => {
+        deletedXML.push(
+            '<Deleted>',
+            '<Key>', escapeForXML(objKey), '</Key>',
+            '</Deleted>'
+        );
+    });
+    xml[2] = deletedXML.join('');
+    return xml.join('');
+}
+
+function _parseXml(xmlToParse, next) {
+    return parseString(xmlToParse, (err, result) => {
+        if (err || !result || !result.Delete) {
+            return next(errors.MalformedXML);
+        }
+        const json = result.Delete;
+        // not quiet is the default if nothing specified
+        const quietSetting = json.Quiet && json.Quiet[0] === 'true';
+        // format of json is {"Object":[{"Key":["test1"]},{"Key":["test2"]}]}
+        const objects = json.Object.map(item => item.Key[0]);
+        return next(null, quietSetting, objects);
+    });
+}
+
+/**
+* gets object metadata and deletes object
+* @param {string} bucketName - bucketName
+* @param {boolean} quietSetting - true if xml should just include error list
+* and false if should include deleted list and error list
+* @param {object []} errorResults - list of error result objects with each
+* object containing -- key: objectName, error: arsenal error
+* @param {string []} inPlay - list of object keys still in play
+* @param {object} log - logger object
+* @param {function} next - callback to next step in waterfall
+* @return {undefined}
+* @callback called with (err, quietSetting, errorResults, numOfObjects,
+* successfullyDeleted, totalContentLengthDeleted)
+*/
+export function getObjMetadataAndDelete(bucketName, quietSetting,
+    errorResults, inPlay, log, next) {
+    const successfullyDeleted = [];
+    let totalContentLengthDeleted = 0;
+    // for obj deletes, no need to check acl's at object level
+    // (authority is at the bucket level for obj deletes)
+
+    // doing 5 requests at a time. note that the data wrapper
+    // will do 5 parallel requests to data backend to delete parts
+    return async.forEachLimit(inPlay, 5, (key, moveOn) => {
+        metadata.getObjectMD(bucketName, key, log, (err, objMD) => {
+            // if general error from metadata return error
+            if (err && !err.NoSuchKey) {
```

Comment:
Good call on the log.  As discussed, I opted to give the user a view of what happened rather than doing a blanket error for all objects.  Not a slam dunk choice but seems better to let them know there were some successes.